doc/man/sconsign.1
.\" __COPYRIGHT__
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining
.\" a copy of this software and associated documentation files (the
.\" "Software"), to deal in the Software without restriction, including
.\" without limitation the rights to use, copy, modify, merge, publish,
.\" distribute, sublicense, and/or sell copies of the Software, and to
.\" permit persons to whom the Software is furnished to do so, subject to
.\" the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included
.\" in all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
.\" KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
.\" WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
.\" NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
.\" LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
.\" OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
.\" WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
.\"
.\" __FILE__ __REVISION__ __DATE__ __DEVELOPER__
.\"
.\" ES - Example Start - indents and turns off line fill
.de ES
.RS
.nf
..
.\" EE - Example End - ends indent and turns line fill back on
.de EE
.RE
.fi
..
.TH SCONSIGN 1 "March 2006"
.SH NAME
sconsign \- print SCons .sconsign file information
.SH SYNOPSIS
.B sconsign
[
.IR options ...
]
.IR file
[ ... ]
.SH DESCRIPTION

The 
.B sconsign
command
displays the contents of one or more
.B .sconsign
files specified by the user.

By default,
.B sconsign
dumps the entire contents of the
specified file(s).
Each entry is printed in the following format:

    file: signature timestamp length
            implicit_dependency_1: signature timestamp length
            implicit_dependency_2: signature timestamp length
            action_signature [action string]

.B None
is printed
in place of any missing timestamp, bsig, or csig
values for
any entry
or any of its dependencies.
If the entry has no implicit dependencies,
or no build action,
the lines are simply omitted.

By default,
.B sconsign
assumes that any
.I file
arguments that end with a
.B .dbm
suffix contains
signature entries for
more than one directory
(that is,
was specified by the
.B SConsignFile ()
function).
Any
.I file
argument that does not end in
.B .dbm
is assumed to be a traditional
.B .sconsign
file containing the signature entries
for a single directory.
An explicit format
may be specified using the
.B -f
or
.B --file=
options.

.SH OPTIONS

Various options control what information is printed
and the format:

.TP
-a, --act, --action
Prints the build action information
for all entries or the specified entries.

.TP
-c, --csig
Prints the content signature (csig) information
for all entries or the specified entries.

.TP
-d DIRECTORY, --dir=DIRECTORY
When the signatures are being
read from a
.B .dbm
file, or the
.B -f dbm
or
.B --format=dbm
options are used,
prints information about
only the signatures
for entries in the specified
.IR DIRECTORY .

.TP
-e ENTRY, --entry=ENTRY
Prints information about only the specified
.IR ENTRY .
Multiple -e options may be used,
in which case information about each
.I ENTRY
is printed in the order in which the
options are specified on the command line.

.TP
-f FORMAT, --format=FORMAT
The file(s) to be printed
are in the specified
.IR FORMAT .
Legal values are
.B dbm
(the DBM format used
when the
.BR SConsignFile ()
function is used)
or
.B sconsign
(the default format
used for an individual
.B .sconsign
file in each directory).

.TP
-h, --help
Prints a help message and exits.

.TP
-i, --implicit
Prints the list of cached implicit dependencies
for all entries or the the specified entries.

.TP
--raw
Prints a pretty-printed representation
of the raw Python dictionary that holds
build information about individual entry
(both the entry itself or its implicit dependencies).
An entry's build action is still printed in its usual format.

.TP
-r, --readable
Prints timestamps in a human-readable string,
enclosed in single quotes.

.TP
-t, --timestamp
Prints the timestamp information
for all entries or the specified entries.

.TP
-v, --verbose
Prints labels identifying each field being printed.

.SH ENVIRONMENT

.IP SCONS_LIB_DIR
Specifies the directory that contains the SCons Python module directory
(e.g. /home/aroach/scons-src-0.01/src/engine).
on the command line.

.SH "SEE ALSO"
.BR scons ,
.B scons
User Manual,
.B scons
Design Document,
.B scons
source code.

.SH AUTHORS
Steven Knight <knight at baldmt dot com>


src/CHANGES.txt
# __COPYRIGHT__
# __FILE__ __REVISION__ __DATE__ __DEVELOPER__


                 SCons - a software construction tool

                            Change Log



RELEASE 0.97 - XXX

  From Ken Boortz:

  - Enhance ParseConfig() to recognize options that begin with '+'.

  From Christopher Drexler:

  - Make SCons aware bibtex must be called if any \include files
    cause creation of a bibliography.

  - Make SCons aware that "\bilbiography" in TeX source files means
    that related .bbl and .blg bibliography files will be created.
    (NOTE:  This still needs to search for the string in \include files.)

  From David Gruener:

  - Fix inconsistent handling of Action strfunction arguments.

  - Preserve white space in display Action strfunction strings.

  From Steven Knight:

  - Speed up the Taskmaster significantly by avoiding unnecessary
    re-scans of Nodes to find out if there's work to be done, having it
    track the currently-executed top-level target directly and not
    through its presence on the target list, and eliminating some other
    minor list(s), method(s) and manipulation.

  - Fix the expansion of $TARGET and $SOURCE in the expansion of
    $INSTALLSTR displayed for non-environment calls to InstallAs().

  - Fix the ability to have an Alias() call refer to a directory
    name that's not identified as a directory until later.

  From Sanjoy Mahajan:

  - Change use of $SOURCES to $SOURCE in all TeX-related Tool modules.

  From Joel B. Mohler:

  - Make SCons aware that "\makeindex" in TeX source files means that
    related .ilg, .ind and .idx index files will be created.
    (NOTE:  This still needs to search for the string in \include files.)

  - Prevent scanning the TeX .aux file for additional files from
    trying to remove it twice when the -c option is used.

  From Greg Noel:

  - Add an env.ParseFlags() method that provides separate logic for
    parsing GNU tool chain flags into a dictionary.
    
  - Add an env.MergeFlags() method to apply an arbitrary dictionary
    of flags to a construction environment's variables.

  From Gary Oberbrunner:

  - Fix parsing tripartite Intel C compiler version numbers on Linux.

  - Extend the ParseConfig() function to recognize -arch and
    -isysroot options.

  - Have the error message list the known suffixes when a Builder call
    can't build a source file with an unknown suffix.

  From Karol Pietrzak:

  - Avoid recursive calls to main() in the program snippet used by the
    SConf subsystem to test linking against libraries.  This changes the
    default behavior of CheckLib() and CheckLibWithHeader() to print
    "Checking for C library foo..." instead of "Checking for main()
    in C library foo...".

  From John Pye:

  - Throw an exception if a command called by ParseConfig() or
    ParseFlags() returns an error.

  From Stefan Seefeld:

  - Initial infrastructure for running SCons tests under QMTest.

  From Atul Varma:

  - Fix detection of Visual C++ Express Edition.



RELEASE 0.96.92 - Mon, 10 Apr 2006 21:08:22 -0400

  NOTE:  This was a pre-release of 0.97 for testing purposes.

  From Anonymous:

  - Fix the intelc.py Tool module to not throw an exception if the
    only installed version is something other than ia32.

  - Set $CCVERSION when using gcc.

  From Matthias:

  - Support generating project and solution files for Microsoft
    Visual Studio version 8.

  - Support generating more than one project file for a Microsoft
    Visual Studio solution file.

  - Add support for a support "runfile" parameter to Microsoft
    Visual Studio project file creation.

  - Put the project GUID, not the solution GUID, in the right spot
    in the solution file.

  From Erling Andersen:

  - Fix interpretation of Node.FS objects wrapped in Proxy instances,
    allowing expansion of things like ${File(TARGET)} in command lines.

  From Stanislav Baranov:

  - Add a separate MSVSSolution() Builder, with support for the
    following new construction variables: $MSVSBUILDCOM, $MSVSCLEANCOM,
    $MSVSENCODING, $MSVSREBUILDCOM, $MSVSSCONS, $MSVSSCONSCOM,
    $MSVSSCONSFLAGS, $MSVSSCONSCRIPT and $MSVSSOLUTIONCOM.

  From Ralph W. Grosse-Kunstleve and Patrick Mezard:

  - Remove unneceesary (and incorrect) SCons.Util strings on some function
    calls in SCons.Util.

  From Bob Halley:

  - Fix C/C++ compiler selection on AIX to not always use the external $CC
    environment variable.

  From August Hörandl:

  - Add a scanner for \include and \import files, with support for
    searching a directory list in $TEXINPUTS (imported from the external
    environment).

  - Support $MAKEINDEX, $MAKEINDEXCOM, $MAKEINDEXCOMSTR and
    $MAKEINDEXFLAGS for generating indices from .idx files.

  From Steven Johnson:

  - Add a NoClean() Environment method and function to override removal
    of targets during a -c clean, including documentation and tests.

  From Steven Knight:

  - Check for whether files exist on disk by listing the directory
    contents, not calling os.path.exists() file by file.  This is
    somewhat more efficient in general, and may be significantly
    more efficient on Windows.

  - Minor speedups in the internal is_Dict(), is_List() and is_String()
    functions.

  - Fix a signature refactoring bug that caused Qt header files to
    get re-generated every time.

  - Don't fail when writing signatures if the .sconsign.dblite file is
    owned by a different user (e.g. root) from a previous run.

  - When deleting variables from stacked OverrideEnvironments, don't
    throw a KeyError if we were able to delte the variable from any
    Environment in the stack.

  - Get rid of the last indentation tabs in the SCons source files and
    add -tt to the Python invocations in the packaging build and the
    tests so they don't creep back in.

  - In Visual Studio project files, put quotes around the -C directory
    so everything works even if the path has spaces in it.


src/script/sconsign.py
#! /usr/bin/env python
#
# SCons - a Software Constructor
#
# __COPYRIGHT__
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
# KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

__revision__ = "__FILE__ __REVISION__ __DATE__ __DEVELOPER__"

__version__ = "__VERSION__"

__build__ = "__BUILD__"

__buildsys__ = "__BUILDSYS__"

__date__ = "__DATE__"

__developer__ = "__DEVELOPER__"

import os
import os.path
import sys
import time

##############################################################################
# BEGIN STANDARD SCons SCRIPT HEADER
#
# This is the cut-and-paste logic so that a self-contained script can
# interoperate correctly with different SCons versions and installation
# locations for the engine.  If you modify anything in this section, you
# should also change other scripts that use this same header.
##############################################################################

# Strip the script directory from sys.path() so on case-insensitive
# (WIN32) systems Python doesn't think that the "scons" script is the
# "SCons" package.  Replace it with our own library directories
# (version-specific first, in case they installed by hand there,
# followed by generic) so we pick up the right version of the build
# engine modules if they're in either directory.

script_dir = sys.path[0]

if script_dir in sys.path:
    sys.path.remove(script_dir)

libs = []

if os.environ.has_key("SCONS_LIB_DIR"):
    libs.append(os.environ["SCONS_LIB_DIR"])

local = 'scons-local-' + __version__
if script_dir:
    local = os.path.join(script_dir, local)
libs.append(local)

scons_version = 'scons-%s' % __version__

prefs = []

if sys.platform == 'win32':
    # sys.prefix is (likely) C:\Python*;
    # check only C:\Python*.
    prefs.append(sys.prefix)
    prefs.append(os.path.join(sys.prefix, 'Lib', 'site-packages'))
else:
    # On other (POSIX) platforms, things are more complicated due to
    # the variety of path names and library locations.  Try to be smart
    # about it.
    if script_dir == 'bin':
        # script_dir is `pwd`/bin;
        # check `pwd`/lib/scons*.
        prefs.append(os.getcwd())
    else:
        if script_dir == '.' or script_dir == '':
            script_dir = os.getcwd()
        head, tail = os.path.split(script_dir)
        if tail == "bin":
            # script_dir is /foo/bin;
            # check /foo/lib/scons*.
            prefs.append(head)

    head, tail = os.path.split(sys.prefix)
    if tail == "usr":
        # sys.prefix is /foo/usr;
        # check /foo/usr/lib/scons* first,
        # then /foo/usr/local/lib/scons*.
        prefs.append(sys.prefix)
        prefs.append(os.path.join(sys.prefix, "local"))
    elif tail == "local":
        h, t = os.path.split(head)
        if t == "usr":
            # sys.prefix is /foo/usr/local;
            # check /foo/usr/local/lib/scons* first,
            # then /foo/usr/lib/scons*.
            prefs.append(sys.prefix)
            prefs.append(head)
        else:
            # sys.prefix is /foo/local;
            # check only /foo/local/lib/scons*.
            prefs.append(sys.prefix)
    else:
        # sys.prefix is /foo (ends in neither /usr or /local);
        # check only /foo/lib/scons*.
        prefs.append(sys.prefix)

    temp = map(lambda x: os.path.join(x, 'lib'), prefs)
    temp.extend(map(lambda x: os.path.join(x,
                                           'lib',
                                           'python' + sys.version[:3],
                                           'site-packages'),
                           prefs))
    prefs = temp

    # Add the parent directory of the current python's library to the
    # preferences.  On SuSE-91/AMD64, for example, this is /usr/lib64,
    # not /usr/lib.
    try:
        libpath = os.__file__
    except AttributeError:
        pass
    else:
        while libpath:
            libpath, tail = os.path.split(libpath)
            if tail[:6] == "python":
                break
        if libpath:
            # Python library is in /usr/libfoo/python*;
            # check /usr/libfoo/scons*.
            prefs.append(libpath)

# Look first for 'scons-__version__' in all of our preference libs,
# then for 'scons'.
libs.extend(map(lambda x: os.path.join(x, scons_version), prefs))
libs.extend(map(lambda x: os.path.join(x, 'scons'), prefs))

sys.path = libs + sys.path

##############################################################################
# END STANDARD SCons SCRIPT HEADER
##############################################################################

import cPickle
import imp
import string
import whichdb

import SCons.SConsign

def my_whichdb(filename):
    try:
        f = open(filename + ".dblite", "rb")
        f.close()
        return "SCons.dblite"
    except IOError:
        pass
    return _orig_whichdb(filename)

_orig_whichdb = whichdb.whichdb
whichdb.whichdb = my_whichdb

def my_import(mname):
    if '.' in mname:
        i = string.rfind(mname, '.')
        parent = my_import(mname[:i])
        fp, pathname, description = imp.find_module(mname[i+1:],
                                                    parent.__path__)
    else:
        fp, pathname, description = imp.find_module(mname)
    return imp.load_module(mname, fp, pathname, description)

class Flagger:
    default_value = 1
    def __setitem__(self, item, value):
        self.__dict__[item] = value
        self.default_value = 0
    def __getitem__(self, item):
        return self.__dict__.get(item, self.default_value)

Do_Call = None
Print_Directories = []
Print_Entries = []
Print_Flags = Flagger()
Verbose = 0
Readable = 0

def default_mapper(entry, name):
    try:
        val = eval("entry."+name)
    except:
        val = None
    return str(val)

def map_action(entry, name):
    try:
        bact = entry.bact
        bactsig = entry.bactsig
    except AttributeError:
        return None
    return '%s [%s]' % (bactsig, bact)

def map_timestamp(entry, name):
    try:
        timestamp = entry.timestamp
    except AttributeError:
        timestamp = None
    if Readable and timestamp:
        return "'" + time.ctime(timestamp) + "'"
    else:
        return str(timestamp)

def map_bkids(entry, name):
    try:
        bkids = entry.bsources + entry.bdepends + entry.bimplicit
        bkidsigs = entry.bsourcesigs + entry.bdependsigs + entry.bimplicitsigs
    except AttributeError:
        return None
    result = []
    for i in xrange(len(bkids)):
        result.append(nodeinfo_string(bkids[i], bkidsigs[i], "        "))
    if result == []:
        return None
    return string.join(result, "\n        ")

map_field = {
    'action'    : map_action,
    'timestamp' : map_timestamp,
    'bkids'     : map_bkids,
}

map_name = {
    'implicit'  : 'bkids',
}

def field(name, entry, verbose=Verbose):
    if not Print_Flags[name]:
        return None
    fieldname = map_name.get(name, name)
    mapper = map_field.get(fieldname, default_mapper)
    val = mapper(entry, name)
    if verbose:
        val = name + ": " + val
    return val

def nodeinfo_raw(name, ninfo, prefix=""):
    # This just formats the dictionary, which we would normally use str()
    # to do, except that we want the keys sorted for deterministic output.
    d = ninfo.__dict__
    try:
        keys = ninfo.field_list
    except AttributeError:
        keys = d.keys()
        keys.sort()
    l = []
    for k in keys:
        l.append('%s: %s' % (repr(k), repr(d.get(k))))
    return name + ': {' + string.join(l, ', ') + '}'

def nodeinfo_cooked(name, ninfo, prefix=""):
    field_list = ['csig', 'timestamp', 'size']
    f = lambda x, ni=ninfo, v=Verbose: field(x, ni, v)
    outlist = [name+':'] + filter(None, map(f, field_list))
    if Verbose:
        sep = '\n    ' + prefix
    else:
        sep = ' '
    return string.join(outlist, sep)

nodeinfo_string = nodeinfo_cooked

def printfield(name, entry, prefix=""):
    print nodeinfo_string(name, entry.ninfo, prefix)

    outlist = field("implicit", entry, 0)
    if outlist:
        if Verbose:
            print "    implicit:"
        print "        " + outlist
    outact = field("action", entry, 0)
    if outact:
        if Verbose:
            print "    action: " + outact
        else:
            print "        " + outact

def printentries(entries):
    if Print_Entries:
        for name in Print_Entries:
            try:
                entry = entries[name]
            except KeyError:
                sys.stderr.write("sconsign: no entry `%s' in `%s'\n" % (name, args[0]))
            else:
                printfield(name, entry)
    else:
        names = entries.keys()
        names.sort()
        for name in names:
            printfield(name, entries[name])

class Do_SConsignDB:
    def __init__(self, dbm_name, dbm):
        self.dbm_name = dbm_name
        self.dbm = dbm

    def __call__(self, fname):
        # The *dbm modules stick their own file suffixes on the names
        # that are passed in.  This is causes us to jump through some
        # hoops here to be able to allow the user
        try:
            # Try opening the specified file name.  Example:
            #   SPECIFIED                  OPENED BY self.dbm.open()
            #   ---------                  -------------------------
            #   .sconsign               => .sconsign.dblite
            #   .sconsign.dblite        => .sconsign.dblite.dblite
            db = self.dbm.open(fname, "r")
        except (IOError, OSError), e:
            print_e = e
            try:
                # That didn't work, so try opening the base name,
                # so that if the actually passed in 'sconsign.dblite'
                # (for example), the dbm module will put the suffix back
                # on for us and open it anyway.
                db = self.dbm.open(os.path.splitext(fname)[0], "r")
            except (IOError, OSError):
                # That didn't work either.  See if the file name
                # they specified just exists (independent of the dbm
                # suffix-mangling).
                try:
                    open(fname, "r")
                except (IOError, OSError), e:
                    # Nope, that file doesn't even exist, so report that
                    # fact back.
                    print_e = e
                sys.stderr.write("sconsign: %s\n" % (print_e))
                return
        except KeyboardInterrupt:
            raise
        except cPickle.UnpicklingError:
            sys.stderr.write("sconsign: ignoring invalid `%s' file `%s'\n" % (self.dbm_name, fname))
            return
        except Exception, e:
            sys.stderr.write("sconsign: ignoring invalid `%s' file `%s': %s\n" % (self.dbm_name, fname, e))
            return

        if Print_Directories:
            for dir in Print_Directories:
                try:
                    val = db[dir]
                except KeyError:
                    sys.stderr.write("sconsign: no dir `%s' in `%s'\n" % (dir, args[0]))
                else:
                    self.printentries(dir, val)
        else:
            keys = db.keys()
            keys.sort()
            for dir in keys:
                self.printentries(dir, db[dir])

    def printentries(self, dir, val):
        print '=== ' + dir + ':'
        printentries(cPickle.loads(val))

def Do_SConsignDir(name):
    try:
        fp = open(name, 'rb')
    except (IOError, OSError), e:
        sys.stderr.write("sconsign: %s\n" % (e))
        return
    try:
        sconsign = SCons.SConsign.Dir(fp)
    except KeyboardInterrupt:
        raise
    except cPickle.UnpicklingError:
        sys.stderr.write("sconsign: ignoring invalid .sconsign file `%s'\n" % (name))
        return
    except Exception, e:
        sys.stderr.write("sconsign: ignoring invalid .sconsign file `%s': %s\n" % (name, e))
        return
    printentries(sconsign.entries)

##############################################################################

import getopt

helpstr = """\
Usage: sconsign [OPTIONS] FILE [...]
Options:
  -a, --act, --action         Print build action information.
  -c, --csig                  Print content signature information.
  -d DIR, --dir=DIR           Print only info about DIR.
  -e ENTRY, --entry=ENTRY     Print only info about ENTRY.
  -f FORMAT, --format=FORMAT  FILE is in the specified FORMAT.
  -h, --help                  Print this message and exit.
  -i, --implicit              Print implicit dependency information.
  -r, --readable              Print timestamps in human-readable form.
  --raw                       Print raw Python object representations.
  -s, --size                  Print file sizes.
  -t, --timestamp             Print timestamp information.
  -v, --verbose               Verbose, describe each field.
"""

opts, args = getopt.getopt(sys.argv[1:], "abcd:e:f:hirstv",
                            ['act', 'action',
                             'csig', 'dir=', 'entry=',
                             'format=', 'help', 'implicit',
                             'raw', 'readable',
                             'size', 'timestamp', 'verbose'])


for o, a in opts:
    if o in ('-a', '--act', '--action'):
        Print_Flags['action'] = 1
    elif o in ('-c', '--csig'):
        Print_Flags['csig'] = 1
    elif o in ('-d', '--dir'):
        Print_Directories.append(a)
    elif o in ('-e', '--entry'):
        Print_Entries.append(a)
    elif o in ('-f', '--format'):
        Module_Map = {'dblite'   : 'SCons.dblite',
                      'sconsign' : None}
        dbm_name = Module_Map.get(a, a)
        if dbm_name:
            try:
                dbm = my_import(dbm_name)
            except:
                sys.stderr.write("sconsign: illegal file format `%s'\n" % a)
                print helpstr
                sys.exit(2)
            Do_Call = Do_SConsignDB(a, dbm)
        else:
            Do_Call = Do_SConsignDir
    elif o in ('-h', '--help'):
        print helpstr
        sys.exit(0)
    elif o in ('-i', '--implicit'):
        Print_Flags['implicit'] = 1
    elif o in ('--raw',):
        nodeinfo_string = nodeinfo_raw
    elif o in ('-r', '--readable'):
        Readable = 1
    elif o in ('-s', '--size'):
        Print_Flags['size'] = 1
    elif o in ('-t', '--timestamp'):
        Print_Flags['timestamp'] = 1
    elif o in ('-v', '--verbose'):
        Verbose = 1

if Do_Call:
    for a in args:
        Do_Call(a)
else:
    for a in args:
        dbm_name = whichdb.whichdb(a)
        if dbm_name:
            Map_Module = {'SCons.dblite' : 'dblite'}
            dbm = my_import(dbm_name)
            Do_SConsignDB(Map_Module.get(dbm_name, dbm_name), dbm)(a)
        else:
            Do_SConsignDir(a)

sys.exit(0)
